File: py_solver/grid.py
"""
Tile predicates and the ASCII tile alphabet.

- `count_grass_tiles` / `is_island_tile` -- used by the solver and the
  renderer. Their TS equivalents live in `lib/solver/island.ts`.
- `TILE_CHAR_MAP` -- the alphabet tests are written in, via
  `tests.helpers.make_grid(["GGG", "G.G"])`, and the character table
  `parity/export_fixtures.py` writes its fixtures with.

The wire format is a blueprint code -- see `blueprint.py` and its peer
`lib/encoding/blueprint.ts`. Nothing here decodes one.
"""

from typing import Dict, List, Optional

from solver.types import Tile

# Tile character -> tile type. Only "grass" is buildable; every other type is
# scenery that blocks placement.
TILE_CHAR_MAP: Dict[str, str] = {
    ".": "water",
    "G": "grass",
    "R": "rock",
    "T": "tree1",
    "U": "tree2",
    "O": "pond",
    "X": "transformer",
}


def is_island_tile(tile: Optional[Tile]) -> bool:
    """True for any land tile (buildable or scenery), i.e. anything but water."""
    return tile is not None and tile.type != "water"


def count_grass_tiles(grid: List[List[Tile]]) -> int:
    return sum(1 for row in grid for tile in row if tile.type == "grass")

File: packages/solver/src/grid.ts
/**
 * Tile predicates and the ASCII tile alphabet.
 *
 * `isIslandTile` is the renderer's and the decomposer's shared question — the
 * solver's own copy of the buildable test lives in `solver/island.ts`, where it
 * is on the hot path. `TILE_CHAR_MAP` is the alphabet the tests are written in,
 * via `makeGrid(["GGG", "G.G"])`.
 *
 * The wire format is a blueprint code — see `encoding/blueprint.ts`. Nothing
 * here decodes one.
 */
import type { Tile, TileType } from "./solver/types";

/**
 * Tile character → tile type. Only `grass` is buildable; every other type is
 * scenery that blocks placement.
 */
export const TILE_CHAR_MAP: Readonly<Record<string, TileType>> = {
  ".": "water",
  G: "grass",
  R: "rock",
  T: "tree1",
  U: "tree2",
  O: "pond",
  X: "transformer",
};

/** True for any land tile (buildable or scenery), i.e. anything but water. */
export function isIslandTile(tile: Tile | null | undefined): boolean {
  return tile != null && tile.type !== "water";
}

/**
 * Builds a tile grid from an ASCII picture, one string per row (y).
 *
 *     makeGrid(["GGG", "G.G"])
 *
 * Exported rather than kept in the test helpers because the CLI reads boards
 * the same way, and an alphabet with two readings is an alphabet with a bug.
 */
export function makeGrid(rows: readonly string[]): Tile[][] {
  const width = rows.length > 0 ? rows[0].length : 0;
  if (!rows.every((r) => r.length === width)) {
    throw new Error("makeGrid rows must be equal length");
  }
  return rows.map((row, y) =>
    [...row].map((char, x): Tile => {
      const type = TILE_CHAR_MAP[char];
      if (type === undefined) {
        throw new Error(`makeGrid: '${char}' is not a tile character`);
      }
      return { x, y, type };
    }),
  );
}
